Catlearn 38 fix random walk edges

Previsou opened the pull request.

Previsou committed:
catlearn-38-fix catlearn-38 Enhance and fix random_walk_edge_sample

Previsou committed:
catlearn-38-fix catlearn-38 Add more precise unit tests for random_walk_edge_sample function

Previsou commented on catlearn/graph_utils.py:
```diff
@@ -737,6 +738,7 @@ def random_walk_edge_sample(
     - n_seeds: to start random walk on specified number of uniformly selected vertices
     - use_opposite: if True, randomly walk either direct or dual graph fairly
     - use_both_ends: if True, randomly select either end of sampled edges for growing the walk
+    - max_degree: If strictly positive, the maximum allowed degree (in and out) of the subgraph
```
Indeed, will update the docstring

Previsou commented on catlearn/graph_utils.py:
```diff
@@ -750,6 +752,10 @@ def random_walk_edge_sample(
     """
+    def get_degree_safe(graph, node, in_degree):
+        degrees = graph.in_degree([node]) if in_degree else graph.out_degree([node])
+        return dict(degrees).get(node, 0)  # thank you networkx...
```
The comment is somewhat sarcastic, but only as a reference to networkkx quirky API on this usage. If I pass a single vertex, I have either a int if the node exstis, or a class otherwise. And then the resulting `DegreeView` has dict-like access but no method to return a default value. I can add extra comments for clarification tough.

I could have used exceptions, but I am personally biased against them, even though they are native python constructs x)

Previsou replied:
Updated

Previsou commented on catlearn/graph_utils.py:
```diff
@@ -766,16 +772,16 @@ def random_walk_edge_sample(
+                (get_degree_safe(sampled_graph, src, in_degree=False) < max_degree
+                 and get_degree_safe(sampled_graph, dst, in_degree=True) < max_degree)):
+                sampled_edges.append((src, dst))
+                sampled_graph.add_edge(src, dst, **labels)
```
why?

igorgarbuz replied:
does it work for you with current implementation ? I copied the code in jupyter, try to run and get an error. Also networkx doc specifies usage for update() to add labels.

Previsou replied:
It worked with the test suite, but I guess they didn't contain any label. Anyhow, even if the current code was working, it does not make explicit what happens when an edge is effectively updated (oerwritten, merged, etlse?). So I'll amend with your proposal.

Previsou replied:
Done

Previsou commented on catlearn/graph_utils.py:
```diff
@@ -721,7 +721,8 @@ def random_walk_edge_sample(
         use_opposite: bool = False,
-        use_both_ends: bool = False) -> DirectedGraph[NodeType]:
+        use_both_ends: bool = False,
+        max_degree: int = 0) -> DirectedGraph[NodeType]:
```
Well this is a *generic* sampler, so it has no idea which labels are 'opposite', and whether it should enforce this external constraints. If the input graph has this property, what prevents you from augmenting the sampled subgraph as a post -processing? Even more so relevant when we do this augmentation ourselves on the full graph.

Previsou commented on catlearn/graph_utils.py:
```diff
@@ -766,16 +772,16 @@ def random_walk_edge_sample(
+                 and get_degree_safe(sampled_graph, dst, in_degree=True) < max_degree)):
+                sampled_edges.append((src, dst))
+                sampled_graph.add_edge(src, dst, **labels)
     return sampled_graph
```
I can't select only a subset of labels by using networkx's subgraph, so that defeat slightly the purpose of this sampler imho

Previsou committed:
catlearn-38-fix catlearn-38 Fix and enhance random_walk_edge_sample function

1. Several refactoring as per the code review feedback:
   optimization, clearer docstring, fixing graph with labels.
2. Separate in- and out- degree constraints

igorgarbuz commented on catlearn/graph_utils.py:
```diff
@@ -845,7 +863,6 @@ def clean_isolates(
     """Clean nodes with zero connections (either in- or outbound)."""
     isolates = list(nx.isolates(graph))
     print(f'Following {len(isolates)} isolate(s) are found:\n{isolates}')
-    graph.remove_nodes_from(list(nx.isolates(graph)))
```
nice catch, thanks.

igorgarbuz commented:
Few comment, but good for me.

Previsou commented on catlearn/graph_utils.py:
```diff
@@ -738,10 +737,17 @@ def random_walk_edge_sample(
 
+    Note: if the seeds do not repsect the degree constraints, they will be kept exactly has_edge
+          passed. Sampling will enforce the constraints when adding edges, but the resulting
+          subgraph will not have the expected capped degree.
```
If and only if the passed seeds already have this property. You could decide to have `max_in_degree=1` and pass `seeds=[1->3, 2->3]` - so not respecting the constraints. In this case the added edges will respect the degree constraints, altough the resulting subgraph won't

Previsou commented on catlearn/graph_utils.py:
```diff
@@ -766,16 +772,16 @@ def random_walk_edge_sample(
         src = e[0] if use_src else e[1]
```
Done

Previsou commented on catlearn/graph_utils.py:
```diff
@@ -738,10 +737,17 @@ def random_walk_edge_sample(
 
     Returns: a valid subgraph
 
+    Note: if the seeds do not repsect the degree constraints, they will be kept exactly has_edge
```
Docstring amended. Looks like a copy paste in the wrong place
